### PR TITLE
Improve the IDE compatibility

### PR DESCRIPTION
**Pull request by leofeyer**
I have completed the work once started by @tristanlins to make Contao more compatible with PHP IDEs (mainly getting rid of the "variable accessed through magic property" warnings). The old PR is available here: https://github.com/contao/core-bundle/pull/1

@contao/developers Please check whether you have any objections.

### CODE REVIEW

File: system/modules/comments/classes/Comments.php [hunk 1]
```diff
@@ -269,6 +276,11 @@ protected function renderCommentForm(\FrontendTemplate $objTemplate, \stdClass $
 			}
 
 			$arrField['eval']['required'] = $arrField['eval']['mandatory'];
+
+			/**
+			 * @var \Widget $strClass
+			 * @var \Widget $objWidget
```
Comment (aschempp) [inline system/modules/comments/classes/Comments.php]: Is this necessary? Does PhpStorm not notice that `new Widget` is of type widget?

File: system/modules/comments/classes/Comments.php [hunk 2]
```diff
@@ -603,7 +618,7 @@ public static function notifyCommentsSubscribers(\CommentsModel $objComment)
 
 		while ($objNotify->next())
 		{
-			// Don't notify the commentor about his own comment
+			/** @var \CommentsNotifyModel $objNotify */
```
Comment (aschempp) [inline system/modules/comments/classes/Comments.php]: This should not be necessary if `\CommentsNotifyModel::findActiveBySourceAndParent` has a correct docblock?

File: system/modules/core/classes/Ajax.php [hunk 3]
```diff
@@ -195,6 +196,7 @@ public function executePostActions(\DataContainer $dc)
 			exit;
 		}
 
+		/** @var \DataContainer|object $dc */
```
Comment (aschempp) [inline system/modules/core/classes/Ajax.php]: As far as I know, `DataContainer` does not accept arbitrary properties, so this should not be `object` but have `@property` for the available options.

File: system/modules/core/classes/Backend.php [hunk 4]
```diff
@@ -841,7 +848,7 @@ public static function addPageIcon($row, $label, DataContainer $dc=null, $imageA
 		// Add the breadcrumb link
 		$label = '<a href="' . \Controller::addToUrl('node='.$row['id']) . '" title="'.specialchars($GLOBALS['TL_LANG']['MSC']['selectNode']).'">' . $label . '</a>';
 
-		// Return the image
+		/** @var \DataContainer|object $dc */
```
Comment (aschempp) [inline system/modules/core/classes/Backend.php]: This should be in the docblock of the function and not inline code.

File: system/modules/core/classes/DataContainer.php [hunk 5]
```diff
@@ -81,7 +92,7 @@ class DataContainer extends \Backend
 
 	/**
 	 * Active record
-	 * @param object
+	 * @param \Model
```
Comment (aschempp) [inline system/modules/core/classes/DataContainer.php]: The class properties should be `@var` or `@type` and not `@param` afaik

File: system/modules/core/classes/DataContainer.php [hunk 6]
```diff
@@ -755,4 +774,17 @@ protected function generateGlobalButtons()
...
+
+	/**
+	 * Return the name of the current palette
+	 * @return string
+	 */
+	abstract public function getPalette();
+
+	/**
+	 * Save the current value
+	 * @param mixed
```
Comment (aschempp) [inline system/modules/core/classes/DataContainer.php]: Missing variable name (`@param mixed $varValue`)

File: system/modules/core/classes/Theme.php [hunk 7]
```diff
@@ -641,7 +643,7 @@ protected function extractThemeFiles($arrFiles, $arrDbFields)
 	 */
 	public function exportTheme(\DataContainer $dc)
 	{
-		// Get the theme meta data
+		/** @var \DataContainer|object $dc */
```
Comment (aschempp) [inline system/modules/core/classes/Theme.php]: This should be a method property, and it should also not force a type then. That would make it easier to just pass in an object with property `id`.

File: system/modules/core/classes/Theme.php [hunk 8]
```diff
@@ -882,7 +884,7 @@ protected function addTableTlImageSize(\DOMDocument $xml, \DOMElement $tables, \
 		$imageSizeItemTable->setAttribute('name', 'tl_image_size_item');
 		$imageSizeItemTable = $tables->appendChild($imageSizeItemTable);
 
-		// Get all sizes
+		/** @var \Database\Result|object $objTheme */
```
Comment (aschempp) [inline system/modules/core/classes/Theme.php]: Maybe we could use `@return \Database\Result|object` for the execute method?

File: system/modules/core/controllers/BackendMain.php [hunk 9]
```diff
@@ -176,95 +260,4 @@ protected function welcomeScreen()
 
 		return $objTemplate->parse();
 	}
-
-
-	/**
-	 * Output the template file
-	 */
-	protected function output()
```
Comment (aschempp) [inline system/modules/core/controllers/BackendMain.php]: This is not a good idea, I'm pretty sure there are people using `BackendMain` as there parent class to override stuff like that.

File: system/modules/core/elements/ContentAlias.php [hunk 10]
```diff
@@ -39,7 +39,7 @@ public function generate()
 			return '';
 		}
 
-		// Overwrite the ID
+		/** @var \ContentElement $objElement */
```
Comment (aschempp) [inline system/modules/core/elements/ContentAlias.php]: Why is this necessary? `\ContentModel::findByPk` should already know the correct type?

File: system/modules/core/elements/ContentDownloads.php [hunk 11]
```diff
@@ -21,7 +21,7 @@ class ContentDownloads extends \ContentElement
 
 	/**
 	 * Files object
-	 * @var \FilesModel
```
Comment (aschempp) [inline system/modules/core/elements/ContentDownloads.php]: You could actually use `@var \FilesModel[]` here and then `foreach` over the values. Then the IDE would already know the type of each item in the foreach-loop.

File: system/modules/core/forms/FormFileUpload.php [hunk 12]
```diff
@@ -238,6 +250,8 @@ public function validate()
 
 					// Generate the DB entries
 					$strFile = $strUploadFolder . '/' . $file['name'];
+
+					/** @var \FilesModel $objFile */
```
Comment (aschempp) [inline system/modules/core/forms/FormFileUpload.php]: `\FilesModel::findByPath` should already return the correct type.

File: system/modules/core/library/Contao/Folder.php [hunk 13]
```diff
@@ -339,6 +340,7 @@ protected function getHash()
 	{
 		$arrFiles = array();
 
+		/** @var \RecursiveDirectoryIterator $it */
```
Comment (aschempp) [inline system/modules/core/library/Contao/Folder.php]: really?

File: system/modules/core/library/Contao/Image.php [hunk 14]
```diff
@@ -869,20 +869,21 @@ public static function create($image, $size=null)
...
-					 ->setTargetHeight($imageSize->height)
-					 ->setResizeMode($imageSize->resizeMode)
-					 ->setZoomLevel($imageSize->zoom);
+			$imageObj->setTargetWidth($imageSize->width);
+			$imageObj->setTargetHeight($imageSize->height);
+			$imageObj->setResizeMode($imageSize->resizeMode);
+			$imageObj->setZoomLevel($imageSize->zoom);
 		}
 
+		/** @var \FilesModel $fileRecord */
```
Comment (aschempp) [inline system/modules/core/library/Contao/Image.php]: Should not be necessary

File: system/modules/core/library/Contao/Model.php [hunk 15]
```diff
@@ -132,6 +134,8 @@ public function __construct(\Database\Result $objResult=null)
 			foreach ($arrRelated as $key=>$row)
 			{
 				$table = $this->arrRelations[$key]['table'];
+
+				/** @var static $strClass */
 				$strClass = static::getClassFromTable($table);
```
Comment (aschempp) [inline system/modules/core/library/Contao/Model.php]: `getClassFromTable` should already `@return static`

File: system/modules/core/library/Contao/Picture.php [hunk 16]
```diff
@@ -118,6 +120,7 @@ public static function create($file, $size=null)
 			$picture->setImageSizeItems(\ImageSizeItemModel::findVisibleByPid($imageSize->id, array('order'=>'sorting ASC')));
 		}
 
+		/** @var \FilesModel $fileRecord */
```
Comment (aschempp) [inline system/modules/core/library/Contao/Picture.php]: This should not be necessary

File: system/modules/core/library/Contao/String.php [hunk 17]
```diff
@@ -579,6 +579,7 @@ public static function srcToInsertTag($data)
 				continue;
 			}
 
+			/** @var \FilesModel $file */
```
Comment (aschempp) [inline system/modules/core/library/Contao/String.php]: This should not be necessary

File: system/modules/core/models/ContentModel.php [hunk 18]
```diff
@@ -14,6 +14,333 @@
...
+ * @property integer $com_perPage
+ * @property boolean $com_moderate
+ * @property boolean $com_bbcode
+ * @property boolean $com_disableCaptcha
+ * @property boolean $com_requireLogin
+ * @property string  $com_template
+ * @property string  $typePrefix
+ * @property string  $classes
+ *
+ * @method static $this findById()
```
Comment (aschempp) [inline system/modules/core/models/ContentModel.php]: `$this` must be wrong in this case.

File: system/modules/core/modules/Module.php [hunk 19]
```diff
@@ -68,6 +188,7 @@ public function __construct($objModule, $strColumn='main')
 
 		parent::__construct();
 
+		/** @var \ModuleModel $objModule */
```
Comment (aschempp) [inline system/modules/core/modules/Module.php]: Thes supported types should be defined in the method property.

File: system/modules/core/modules/ModuleBooknav.php [hunk 20]
```diff
@@ -154,6 +155,7 @@ protected function getBookPages($intParentId, $groups, $time)
 		{
 			while ($objPages->next())
 			{
+				/** @var \PageModel $objPages */
```
Comment (aschempp) [inline system/modules/core/modules/ModuleBooknav.php]: if `\PageModel::findPublishedSubpagesWithoutGuestsByPid` would `@return \Model\Collection|static[]` and we'd use `foreach` that would not be necessary.

**Comment by leofeyer** [comment]
@aschempp Can you please stop posting "should not be necessary" comments? Every change I have made was necessary, so unless it works in your PhpStorm instance …

File: system/modules/core/modules/ModuleRssReader.php [hunk 21]
```diff
@@ -153,6 +160,7 @@ protected function compile()
 		$items = array();
 		$last = min($limit, count($arrItems)) - 1;
 
+		/** @var \SimplePie_Item[] $arrItems $i */
```
Comment (aschempp) [inline system/modules/core/modules/ModuleRssReader.php]: `$i` should not be here

File: system/modules/comments/classes/Comments.php (hunk 1 above)
Comment (leofeyer) [inline system/modules/comments/classes/Comments.php]: Nope.

File: system/modules/comments/classes/Comments.php (hunk 2 above)
Comment (leofeyer) [inline system/modules/comments/classes/Comments.php]: Yes, it is necessary. The method returns a model collection but the IDE needs to understand that we are accessing the model.

File: system/modules/core/classes/Ajax.php (hunk 3 above)
Comment (leofeyer) [inline system/modules/core/classes/Ajax.php]: True.

File: system/modules/core/pages/PageRegular.php [hunk 22]
```diff
@@ -87,6 +89,7 @@ public function generate($objPage, $blnCheckRequest=false)
 			{
 				while ($objModules->next())
```
Comment (aschempp) [inline system/modules/core/pages/PageRegular.php]: If we use `foreach` on the collection, you could use `@var \ModuleModel[] $objModules`

File: system/modules/core/classes/Backend.php (hunk 4 above)
Comment (leofeyer) [inline system/modules/core/classes/Backend.php]: True.

File: system/modules/core/classes/DataContainer.php (hunk 5 above)
Comment (leofeyer) [inline system/modules/core/classes/DataContainer.php]: You are right. There are several of these wrong usages.

File: system/modules/core/widgets/FileSelector.php [hunk 23]
```diff
@@ -200,9 +208,9 @@ protected function renderFiletree($path, $intMargin, $mount=false, $blnProtected
 		$xtnode = 'tree_' . $this->strTable . '_' . $this->strName;
 
 		// Get session data and toggle nodes
-		if ($this->Input->get($flag.'tg'))
+		if (Input::get($flag.'tg'))
```
Comment (aschempp) [inline system/modules/core/widgets/FileSelector.php]: Missing global namespace prefix.

File: system/modules/core/classes/Theme.php (hunk 7 above)
Comment (leofeyer) [inline system/modules/core/classes/Theme.php]: I'll change this.

File: system/modules/core/classes/Theme.php (hunk 8 above)
Comment (leofeyer) [inline system/modules/core/classes/Theme.php]: I am not changing the type of `$objSizes` but of `$objTheme` :)

Comment (aschempp) [inline system/modules/core/classes/Theme.php]: True, sorry. I think the comment should be where you declare the variable then.

File: system/modules/core/controllers/BackendMain.php (hunk 9 above)
Comment (leofeyer) [inline system/modules/core/controllers/BackendMain.php]: I have reverted this already because it had other side effects, too.

File: system/modules/core/elements/ContentAlias.php (hunk 10 above)
Comment (leofeyer) [inline system/modules/core/elements/ContentAlias.php]: Yes, but `$objElement` is a content model first and then a content element. Ugly, I know :(

File: system/modules/core/forms/FormFileUpload.php (hunk 12 above)
Comment (leofeyer) [inline system/modules/core/forms/FormFileUpload.php]: No, it does not, because `findByPath` returns a collection (the path is not unique). Stumbled upon it as well.

File: system/modules/core/library/Contao/Folder.php (hunk 13 above)
Comment (leofeyer) [inline system/modules/core/library/Contao/Folder.php]: Yep :)

There is another strange PHP behavior going on in the `Theme` class, which I however did not fix: "Method 'getAttribute' not found in class \DOMNode" :)

File: system/modules/core/library/Contao/Model.php (hunk 15 above)
Comment (leofeyer) [inline system/modules/core/library/Contao/Model.php]: No, it actually returns a string.

File: system/modules/core/modules/Module.php (hunk 19 above)
Comment (leofeyer) [inline system/modules/core/modules/Module.php]: We'd still need to tell the IDE that we are using a `ModuleModel` here.

File: system/modules/core/modules/ModuleRssReader.php (hunk 21 above)
Comment (leofeyer) [inline system/modules/core/modules/ModuleRssReader.php]: True.

File: system/modules/core/pages/PageRegular.php (hunk 22 above)
Comment (leofeyer) [inline system/modules/core/pages/PageRegular.php]: You are right and this probably applies to all places where we are traversing a model collection.

File: system/modules/core/elements/ContentAlias.php (hunk 10 above)
Comment (aschempp) [inline system/modules/core/elements/ContentAlias.php]: Aah, that happens later. Maybe move the comment to the place it actually happens…

File: system/modules/core/widgets/FileSelector.php (hunk 23 above)
Comment (leofeyer) [inline system/modules/core/widgets/FileSelector.php]: True.

File: system/modules/core/forms/FormFileUpload.php (hunk 12 above)
Comment (aschempp) [inline system/modules/core/forms/FormFileUpload.php]: Hmm, could we use `@return \Model\Collection|static` for all collections? Because the API for would be identical and available?

**Comment by leofeyer** [comment]
All issues mentioned fixed in 6df04c2b92de3b1f64e612d86916191dc2440f01.

File: system/modules/core/pages/PageRegular.php (hunk 22 above)
Comment (aschempp) [inline system/modules/core/pages/PageRegular.php]: Jup. I stopped using `->next()` because it can lead to issues (due to the internal position pointer). As an example, if you use `getRelated` with a collection as result, each call would get you the same collection with the previous internal pointer (possibly set to the last result, so `next()` would return `false`).

File: system/modules/core/modules/Module.php (hunk 19 above)
Comment (aschempp) [inline system/modules/core/modules/Module.php]: Sorry I meant `@property \ModuleModel|\Model\Collection|\Database\Result $objModule`

File: system/modules/core/library/Contao/Model.php (hunk 15 above)
Comment (aschempp) [inline system/modules/core/library/Contao/Model.php]: True, agreed.

**Comment by leofeyer** [comment]
I am evaluating your proposal in `system/modules/core/pages/PageRegular.php`:

```php
// Create a mapper array in case a module is included more than once (see #4849)
if ($objModules !== null)
{
	/**  @var \ModuleModel[] $objModules */
	foreach ($objModules as $objModule)
	{
		$arrMapper[$objModule->id] = $objModule;
	}

	…
```

This is what you meant, right? There are about 200 places where it would have to be changed.

**Comment by aschempp** [comment]
Yes that's what I mean. Great work so far! 

I dislike the `@method` for magic methods. I personally think people should use `findBy('field', 'value')`, there's no reason for the magic methods…

**Comment by leofeyer** [comment]
I always use them :)

**Comment by aschempp** [comment]
> I always use them :)

I know ^^

**Comment by leofeyer** [comment]
While thinking about this, should we not generally replace every `@return Model\Collection` with `@return Model[]`?

**Comment by aschempp** [comment]
That would mean the Collection methods (like `next` and `reset`) would not be available. But `@return \Model\Collection|static[]` should do it.

**Comment by leofeyer** [comment]
Unfortunately it does not work like this:

```php
// Does not work
/**
 * @return static[]|\Model\Collection|null
 */

// This works though
/**
 * @return \FilesModel[]|\Model\Collection|null
 */
```

**Comment by aschempp** [comment]
what happens if you just use `static`? because the model properties are also available on the model collection?

**Comment by leofeyer** [comment]
I am done. Feel free to check the changes again before I merge the PR.

File: system/modules/core/drivers/DC_Table.php [hunk 24]
```diff
@@ -15,99 +15,102 @@
  * Provide methods to modify the database.
  *
  * @author Leo Feyer <https://github.com/leofeyer>
+ * @author Andreas Schempp <https://github.com/schempp>
```
Comment (aschempp) [inline system/modules/core/drivers/DC_Table.php]: Not sure if I contributed here, but my GitHub user is "aschempp" ;-)

File: system/modules/core/elements/ContentGallery.php [hunk 25]
```diff
@@ -284,12 +286,15 @@ protected function compile()
 			// Do not index or cache the page if the page number is outside the range
 			if ($page < 1 || $page > max(ceil($total/$this->perPage), 1))
 			{
+				/** @var \PageModel $objPage */
```
Comment (aschempp) [inline system/modules/core/elements/ContentGallery.php]: This global statement is actually unnecessary because `$objPage is already defined in the first line of `compile()` method.

File: system/modules/core/library/Contao/Folder.php (hunk 13 above)
Comment (aschempp) [inline system/modules/core/library/Contao/Folder.php]: Aah, got it. But using `@var \RecursiveDirectoryIterator[] $it` and a foreach-loop would be way more obvious ^^

File: system/modules/core/modules/ModuleRegistration.php [hunk 26]
```diff
@@ -14,6 +14,8 @@
 /**
  * Front end module "registration".
  *
+ * @property array $editable
```
Comment (aschempp) [inline system/modules/core/modules/ModuleRegistration.php]: This is unnecessary because parent class `Module` already defines the property

File: system/modules/core/drivers/DC_Table.php (hunk 24 above)
Comment (leofeyer) [inline system/modules/core/drivers/DC_Table.php]: Typo, I'm sorry.

File: system/modules/core/pages/PageError403.php [hunk 27]
```diff
@@ -39,7 +39,7 @@ public function generate($pageId, $objRootPage=null)
 			$objRootPage = \PageModel::findPublishedById(is_integer($objRootPage) ? $objRootPage : $objRootPage->id);
 		}
 
-		// Look for an error_403 page
+		/** @var \PageModel $obj403 */
```
Comment (aschempp) [inline system/modules/core/pages/PageError403.php]: Unnecessary, `find403ByPid` already returns `static`

File: system/modules/core/modules/ModuleRegistration.php (hunk 26 above)
Comment (leofeyer) [inline system/modules/core/modules/ModuleRegistration.php]: True.

File: system/modules/core/pages/PageError404.php [hunk 28]
```diff
@@ -68,7 +69,7 @@ public function generate($pageId, $strDomain=null, $strHost=null, $blnUnusedGet=
 			}
 		}
 
-		// Look for an 404 page
+		/** @var \PageModel $obj404 */
```
Comment (aschempp) [inline system/modules/core/pages/PageError404.php]: Also unnecessary

**Comment by leofeyer** [comment]
The first two issues have been fixed in 0cb856664a133acdca4cd817087c45d6b3551f97. The third issue `\PageModel $obj403` is an issue I have come across quite often:

```php
$obj403 = \PageModel::find403ByPid($objRootPage->id);

// Die if there is no page at all
if ($obj403 === null)		// <-- PhpStorm recognizes the PageModel here
{
	header('HTTP/1.1 403 Forbidden');
	die_nicely('be_forbidden', 'Forbidden');
}

// Generate the error page
if (!$obj403->autoforward || !$obj403->jumpTo)		// <-- PhpStorm no longer recognizes the PageModel here
{
	…
```

This seems to be caused by the strict comparison operator (`===`). If you use the normal operator (`==`) or if you use a yoda condition (`null === $obj403`), everything works as expected. Adding another `/** @var … */` comment also fixes it.

Would you prefer the yoda condition over the comment?

File: system/modules/core/library/Contao/Folder.php (hunk 13 above)
Comment (leofeyer) [inline system/modules/core/library/Contao/Folder.php]: I'll change it.

**Comment by aschempp** [comment]
I would got for yoda then, as that's anyway what we're gonna use in Contao 4

File: system/modules/core/pages/PageRegular.php [hunk 29]
```diff
@@ -232,6 +238,7 @@ protected function getPageLayout($objPage)
 			die_nicely('be_no_layout', 'No layout specified');
 		}
 
+		/** @var \LayoutModel $objLayout */
```
Comment (aschempp) [inline system/modules/core/pages/PageRegular.php]: Yoda here too

File: system/modules/core/pages/PageRoot.php [hunk 30]
```diff
@@ -38,9 +39,12 @@ public function generate($pageId, $blnReturn=false)
 			die_nicely('be_no_active', 'No active pages found');
 		}
 
+		/** @var \PageModel $objNextPage */
```
Comment (aschempp) [inline system/modules/core/pages/PageRoot.php]: Yoda problem again. I'll stop with that, guessing you'll find them ;-)

File: system/modules/news/modules/ModuleNewsArchive.php [hunk 31]
```diff
@@ -161,12 +164,15 @@ protected function compile()
 				// Do not index or cache the page if the page number is outside the range
 				if ($page < 1 || $page > max(ceil($total/$this->perPage), 1))
 				{
+					/** @var \PageModel $objPage */
 					global $objPage;
```
Comment (aschempp) [inline system/modules/news/modules/ModuleNewsArchive.php]: objPage is already set global at the beginning of the method.

File: system/modules/news/modules/ModuleNews.php [hunk 32]
```diff
@@ -350,10 +369,12 @@ protected function generateNewsUrl($objItem, $blnAddArchive=false)
 
 	/**
 	 * Generate a link and return it as string
-	 * @param string
-	 * @param object
-	 * @param boolean
-	 * @param boolean
+	 *
+	 * @param string  $strLink
+	 * @param object  $objArticle
```
Comment (aschempp) [inline system/modules/news/modules/ModuleNews.php]: `$objArticle` should be of type `\NewsModel`.

**Comment by aschempp** [comment]
`Controller::printArticleAsPdf` should expect `\ModuleModel` as argument.

File: system/modules/core/controllers/FrontendIndex.php [hunk 33]
```diff
@@ -56,7 +56,9 @@ public function __construct()
 	 */
 	public function run()
 	{
+		/** @var \PageModel $objPage */
```
Comment (aschempp) [inline system/modules/core/controllers/FrontendIndex.php]: You should remove this as it prevents line 93 from correctly noticing the variable could be a `\Model\Collection`

Comment (leofeyer) [inline system/modules/core/controllers/FrontendIndex.php]: Thanks a lot. I just stumbled across this, too, and I had not yet figured out :)

**Comment by aschempp** [comment]
According to [PhpDoc](http://manual.phpdoc.org/HTMLSmartyConverter/PHP/phpDocumentor/tutorial_tags.property.pkg.html) we sould use `@property-read` and `@property-write` for read-only or write-only properties. For example this affects class `\Contao\Request`, where `code`, `error` etc. are only readable.
It does not make a difference for the IDE though.

Apart from that, I think we're done here. Need to iron some clothes now ^^

**Comment by leofeyer** [comment]
Big thanks for taking the time!

**Comment by leofeyer** [comment]
Merged in 5b8c493d0856bf6038f6e76b92225335a1635952.
